File: boj/b_18870.py
import sys
sys.stdin = open("input.txt", "r") # 제거
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
table, cnt = {}, -1
for i, v in enumerate(sorted(arr)):
    if v not in table:
        cnt += 1
        table[v] = cnt
for a in arr:
    print(table[a])

File: boj/b_18870.js
function solution(n, arr) {
  let arr_s = [...arr].sort((a, b) => a - b);
  let [table, cnt] = [{}, -1];
  for (let i = 0; i < n; i += 1) {
    if (table[arr_s[i]] === undefined) table[arr_s[i]] = ++cnt;
  }
  const result = [];
  arr.forEach((a) => result.push(table[a]));
  return result;
}

console.log(solution(6, [2, 3, 0, 3, 1]));
